joomla/joomla-cms
# [5.0] Add exceptions to deleted file check for plugins which are uninstalled on update by script.php

Pull Request for Issue # .

### Summary of Changes

Sine 5.0.0-beta1, a few more plugins are uninstalled by the [uninstallExtensions method in the script.php file](https://github.com/joomla/joomla-cms/blob/5.0-dev/administrator/components/com_admin/script.php#L273):
- (visible) recaptcha plugin removed, see PR #41563 
- logrotation system plugin replaced by scheduled task, see PR #40519 
- sessiongc system plugin replaced by scheduled task, see PR #41326 
- updatenotification system plugin replaced by scheduled task, see PR #40788 

The files and folders of these plugins shall not be in the lists of files and folders to be deleted on update by the [deleteUnexistingFiles method in script.php](https://github.com/joomla/joomla-cms/blob/5.0-dev/administrator/components/com_admin/script.php#L560), otherwise you will get warning alerts and PHP notices at the end of the update. See PR #41065 for details.

This PR here adds the corresponding files and folders to the lists of exceptions in the "build/deleted_file_check.php" script so they are not falsely reported as to be deleted by that tool, which is used by maintainers and release managers to check for deleted files and folders and by me to maintain the lists in script.php.

In addition, it adds the files and folders from the invisible recaptcha plugin. This one is not uninstalled on update, and its files and folders shall not be deleted, but it's also not included in new releases, so the "build/deleted_file_check.php" script currently reports them as to be deleted.

### Testing Instructions

Code review, or if you want to make a real test:

It needs a current 5.0-dev branch on a git clone and PHP CLI available.

1. Download the [5.0.0-alpha4 full installation package](https://github.com/joomla/joomla-cms/releases/download/5.0.0-alpha4/Joomla_5.0.0-alpha4-Alpha-Full_Package.zip) and unpack it into a folder, e.g. folder "Joomla_5.0.0-alpha4-Alpha-Full_Package" in the "tmp" folder of your Joomla root.
2. Download the [5.0.0-beta1 full installation package](https://github.com/joomla/joomla-cms/releases/download/5.0.0-beta1/Joomla_5.0.0-beta1-Beta-Full_Package.zip) and unpack it into a folder, e.g. folder "Joomla_5.0.0-beta1-Beta-Full_Package" in the "tmp" folder of your Joomla root.
3. In the Joomla root folder in a command shell (e.g. Bash on Linux or CMD on Windows), run the "build/deleted_file_check.php" script to compare the folder from step 1 with the folder from step 2, e.g.:
```
php ./build/deleted_file_check.php --from=./tmp/Joomla_5.0.0-alpha4-Alpha-Full_Package --to=./tmp/Joomla_5.0.0-beta1-Beta-Full_Package
```
4. Check the output of the script on the command line.
5. Save files "build/deleted_files.txt", "build/deleted_folders.txt" and "build/renamed_files.txt" for later comparison.
6. Apply the changes of this PR.
7. Repeat steps 3 to 5.
8. Compare the files from step 5.

### Actual result BEFORE applying this Pull Request

Output of the script:
> There are 32 deleted files, 22 deleted folders and 0 renamed files in comparison to "./tmp/Joomla_5.0.0-alpha4-Alpha-Full_Package"

Files "build/deleted_files.txt" and "build/deleted_folders.txt" contain files and folders from the plugins mentioned in the description of this PR. "build/deleted_folders.txt" contains only such files.

File "build/renamed_files.txt" is empty as it should be.

### Expected result AFTER applying this Pull Request

Output of the script:
> There are 5 deleted files, 0 deleted folders and 0 renamed files in comparison to "./tmp/Joomla_5.0.0-alpha4-Alpha-Full_Package"

Files "build/deleted_files.txt" and "build/deleted_folders.txt" don't contain files and folders from the plugins mentioned in the description of this PR.

See comparison of "build/deleted_files.txt", left side with the PR applied, right side without:
![2023-09-14_deleted-file-check_1](https://github.com/joomla/joomla-cms/assets/7413183/2728f924-53c6-4eff-983c-41486529b599)

File "build/deleted_folders.txt" is empty now as it should be.

File "build/renamed_files.txt" is empty as it should be.

### Link to documentations
Please select:
